### ISSUE TITLE
numba.get_num_threads/set_num_threads resets the value of torch.get_num_threads

### ISSUE THREAD

**Issue by itzsimpl** [numba/numba#9387]
First call of `numba.get_num_threads()` or `numba.get_num_threads()` resets `torch` num_threads value to CPU count. 

Minimal demos:
``` python
$ srun --mem=48G -c32 --container-image=./nemo_head.sqfs --pty bash
root@x:/workspace/nemo# nproc
32
root@x:/workspace/nemo# export OMP_NUM_THREADS=8
root@x:/workspace/nemo# nproc
8
root@x:/workspace/nemo# python3
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import multiprocessing as mp, numba, torch

>>> torch.__version__
'2.1.0a0+29c30b1'

>>> numba.__version__
'0.57.1+1.gc785c8f1f'

>>> mp.cpu_count()
256

>>> torch.get_num_threads()
8

>>> print(torch.__config__.parallel_info())
ATen/Parallel:
        at::get_num_threads() : 8
        at::get_num_interop_threads() : 128
OpenMP 201511 (a.k.a. OpenMP 4.5)
        omp_get_max_threads() : 8
Intel(R) oneAPI Math Kernel Library Version 2021.1-Product Build 20201104 for Intel(R) 64 architecture applications
        mkl_get_max_threads() : 8
Intel(R) MKL-DNN v2.7.3 (Git Hash N/A)
std::thread::hardware_concurrency() : 256
Environment variables:
        OMP_NUM_THREADS : 8
        MKL_NUM_THREADS : [not set]
ATen parallel backend: OpenMP

>>> numba.get_num_threads()
32

>>> torch.get_num_threads()
32

>>> print(torch.__config__.parallel_info())
ATen/Parallel:
        at::get_num_threads() : 32
        at::get_num_interop_threads() : 128
OpenMP 201511 (a.k.a. OpenMP 4.5)
        omp_get_max_threads() : 32
Intel(R) oneAPI Math Kernel Library Version 2021.1-Product Build 20201104 for Intel(R) 64 architecture applications
        mkl_get_max_threads() : 32
Intel(R) MKL-DNN v2.7.3 (Git Hash N/A)
std::thread::hardware_concurrency() : 256
Environment variables:
        OMP_NUM_THREADS : 8
        MKL_NUM_THREADS : [not set]
ATen parallel backend: OpenMP

>>> numba.set_num_threads(2)
>>> numba.get_num_threads()
2

>>> torch.get_num_threads()
32

>>> exit()
root@x:/workspace/nemo# exit
exit
```

```python
root@x:/workspace/nemo# python3
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numba, torch
>>> torch.get_num_threads()
8
>>> torch.set_num_threads(4)
>>> torch.get_num_threads()
4
>>> numba.get_num_threads()
32
>>> torch.get_num_threads()
32
>>> torch.set_num_threads(4)
>>> numba.get_num_threads()
32
>>> torch.get_num_threads()
4
>>> numba.set_num_threads(2)
>>> torch.get_num_threads()
4
>>> numba.get_num_threads()
2
>>> 
```

**Comment by stuartarchibald** [comment]
Thanks for the report @itzsimpl. I've not tried to reproduce as the reported output is sufficiently detailed and it is also what I would guess would happen based on the description.

Some explanation as to what is happening... when a call is made to `numba.get_num_threads()` or `numba.set_num_threads()` if Numba hasn't launched a threading layer already, the first thing it will do is try and launch _a_ threading layer based on availability and user preference (there are 3 to choose from). The threading layer will "read" the number of threads required indirectly from the environment variable `NUMBA_NUM_THREADS` and if that is not present it will typically use `max(1, len(os.sched_getaffinity(0)))`. At this point the threading layer will launch and initialize, in the case of the OpenMP threading layer, as demonstrabily used in the example above, this will include setting the number of threads available to the OpenMP library to `NUMBA_NUM_THREADS` i.e. the OpenMP ICV `nthreads-var` ends up with the value `NUMBA_NUM_THREADS`. It so happens in the example above that `torch` is also using OpenMP and will pick up this change as a result of launching Numba's OpenMP based threading layer (they are clearly sharing the OpenMP ICV as only one OpenMP library is realistically ever going to be in process).

This explanation was discussed at the Numba maintainer's meeting on 2024-01-16 and it wasn't clear what the expectation is for `torch` and Numba interacting. Altering Numba's OpenMP threading layer to check the current thread-pool size, compare that to `NUMBA_NUM_THREADS` or `sched_getaffinity[0]` and then take some action based on it would lead to a deviation in behaviour of the Numba OpenMP threading layer in comparison to the TBB and workqueue threading layers. One thing that might work would be to set `NUMBA_NUM_THREADS` to the same value as whatever is being used to set the number of threads in `torch`, if that's the desired behaviour i.e.
`NUMBA_NUM_THREADS = HOW_TORCH_SETS_MAX_THREADS = <value>`.

It should also be noted that the reason for Numba having its own method for setting a thread count (`NUMBA_NUM_THREADS` env var) is that Numba has three threading layers and for ease of use it was decided that the method of setting a thread count needed to be invariant of the layer used (e.g. `NUMBA_NUM_THREADS` works for all of Numba's threading layers contrary to OpenMP specific env var `OMP_NUM_THREADS` which has no impact on the `TBB` or `workqueue` threading layer).

I hope this explanation helps, please do share your thoughts. Many thanks.

**Comment by itzsimpl** [comment]
@stuartarchibald Thank you for attention and the detailed explanation. Note that I am by no means an OMP expert, I've discovered this issue while debugging a severe degradation in training performance. What worries me in the case of PyTorch and Numba interacting is the interdependency and potential risks that arise from the order of initialisations. In part due to Numba's choice to have its own environment variable independent from OMP, which may, when `NUMBA_NUM_THREADS` is not set, lead to the initialisation of a large pool that is then mask out. 

If PyTorch initialises first, Numba will override the value, which may have a negative impact on PyTorch (CPU oversubscription in a multiprocessing case), but the override will be reflected in PyTorch's `get_num_threads`. However, what will happen if Numba is first, followed by PyTorch? If having a single shared OpenMP ICV is to blame, will a lower or higher value set through PyTorch affect Numba's thread pool as well? And will this be visible at all, since Numba uses masking? Is it safe to assume the entire pool still available? If not, and the choice of next thread is random from the pool, this may result in very hard to debug random crashes.

In view of the last paragraph, is the approach to cache the current torch num threads value prior to checking or updating Numba's num threads, and reapplying torch num threads immediately afterwards, safe or not? Most specifically when `OMP_NUM_THREADS` is set and `NUMBA_NUM_THREADS` is not. 

I do not have any clear idea of any simple or elegant solution that wouldn't require special treatment of the OMP threading layer in Numba. For the time being the best solution seems to be to have both variables set and most importantly also of the same value.

**Comment by stuartarchibald** [comment]
> @stuartarchibald Thank you for attention and the detailed explanation. Note that I am by no means an OMP expert, I've discovered this issue while debugging a severe degradation in training performance. What worries me in the case of PyTorch and Numba interacting is the interdependency and potential risks that arise from the order of initialisations. In part due to Numba's choice to have its own environment variable independent from OMP, which may, when `NUMBA_NUM_THREADS` is not set, lead to the initialisation of a large pool that is then mask out.

No problem RE explanation, that there's a performance degredation in some circumstances when Numba and PyTorch interact is very much an issue of concern. Numba's choice to have its own environment variable is to make its threading backend configuration "portable" across backends. Whether TBB or OpenMP is available is up to the user and in the interests of making it so that setting the number of threads "just works" this was the implementation chosen. Perhaps a solution to this would be to query the size of e.g. OpenMP's threadpool when Numba initialises and if it's something that isn't `NUMBA_NUM_THREADS` or the maximum number of cores, then it must have been altered by some other means and needs to be set to that.

> If PyTorch initialises first, Numba will override the value, which may have a negative impact on PyTorch (CPU oversubscription in a multiprocessing case), but the override will be reflected in PyTorch's `get_num_threads`. However, what will happen if Numba is first, followed by PyTorch? If having a single shared OpenMP ICV is to blame, will a lower or higher value set through PyTorch affect Numba's thread pool as well? And will this be visible at all, since Numba uses masking? Is it safe to assume the entire pool still available? If not, and the choice of next thread is random from the pool, this may result in very hard to debug random crashes.

If Numba is first followed by PyTorch, and PyTorch (or anything else) alters the size of the OpenMP threadpool, then it will potentially cause Numba problems as it assumes that the threadpool is of the size at initialisation so that it can do masking. This is a similar issue to one Numba had with caching of JIT compiled functions whereby  the implementation was accidentally baking the threadpool initialisation size into the cached implementation, which is fine until that size is changed at a later time and the code replayed from cache and there's mismatch (this is now fixed).

> In view of the last paragraph, is the approach to cache the current torch num threads value prior to checking or updating Numba's num threads, and reapplying torch num threads immediately afterwards, safe or not? Most specifically when `OMP_NUM_THREADS` is set and `NUMBA_NUM_THREADS` is not.

Without knowing where Numba initialises the threadpool in the sequence of events it's hard to determine the safety of such an approach, it's at the initialisation of the threadpool that the size is set. In the case of OpenMP, Numba just sets the size of the OpenMP threadpool on initialisation of Numba's parallel backend (by writing to the OpenMP ICV) and it won't read from or write to it again.
 
> I do not have any clear idea of any simple or elegant solution that wouldn't require special treatment of the OMP threading layer in Numba. For the time being the best solution seems to be to have both variables set and most importantly also of the same value.

This is probably the safest thing to do at present.
